stratum/hal/lib/common/p4_service.h

```diff
@@ -232,12 +190,20 @@ class P4Service final : public ::p4::v1::P4Runtime::Service {
   // Channels and threads.
   mutable absl::Mutex stream_response_thread_lock_;
 
-  // Map from node ID to the set of Controller instances corresponding to the
-  // external controller clients connected to that node. The Controller
-  // instances for each node are sorted such that the master (Controller
-  // with highest election_id) is the first element.
-  std::map<uint64, std::set<Controller, ControllerComp>> node_id_to_controllers_
-      GUARDED_BY(controller_lock_);
+  // P4Runtime can accept multiple connections to a single switch for
+  // redundancy. When there is >1 connection the switch chooses a primary which
+  // is used for PacketIO, and is the only connection allowed to write updates.
+  //
+  // It is possible for connections to be made for specific roles. In which case
+  // one primary connection is allowed for each distinct role.
+  std::map<uint64, p4runtime::SdnControllerManager>
+      node_id_to_controller_manager_ ABSL_GUARDED_BY(controller_lock_);
```

Comment:
Check if this can be a absl hash map. We do not depend on the ordering anymore.

Reply:
I encountered problems with the absl flat_hash_map and in-place object creation, it still tried to move the controller. For now this will remain a `std::unordered_map`.